<a href="https://colab.research.google.com/github/shivansh2310/Finance-Stuff/blob/main/Black_Scholes_Model_in_Python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
from scipy.stats import norm

In [4]:
#This function calculates the value of the European option based on Black-Scholes formula

def black_scholes_calc(S0, K, r, T, sigma, option_type):
  d1 = 1/(sigma*np.sqrt(T)) * (np.log(S0/K) + (r+sigma**2/2)*T)
  d2 = d1 - sigma*np.sqrt(T)

  nd1 = norm.cdf(d1)
  nd2 = norm.cdf(d2)

  n_d1 = norm.cdf(-d1)
  n_d2 = norm.cdf(-d2)

  #Put and call Formula

  c = nd1*S0 - nd2*K*np.exp(-r*T)   
  p = K*np.exp(-r*T)*n_d2 - S0*n_d1

# 4) define which value to return based on the option_type parameter
  if option_type=='call':
    return c
  elif option_type=='put':
    return p
  else:
    print('Wrong option type specified')



In [5]:
S0=8.; K=9.; T=3/12.; r=.01; sigma=.2
black_scholes_calc(S0, K, r, T, sigma, 'put')

1.0298160885134449

Running the function gives the result of 1.0298. Recall that holding a put option means that the asset holder has a right to sell the underlying asset at the predefined price (9 in our scenario) at the specified time (in 3 months in our scenario). It is likely that the asset holder expects the price of the asset to decrease. By entering into the put option contract, she secures the opportunity to sell the asset at the price of 9, and pays 1.0298 for the contract itself. 

To make the profit, the asset price at the option expiry should correspond the following rule:

9 – 1.0298 – Price > 0, i.e., Price < 7.97